libenv/sysinfo.c

```diff
@@ -2666,103 +2666,119 @@ static void GetCPUInfo(EvalContext *ctx)
 #ifndef __MINGW32__
 
 /**
-   Return the number of seconds the system has been online given the current
-   time() as an argument, or return -1 if unavailable or unimplemented.
+   Return the current boot time (number of seconds since Epoch).
 */
-int GetUptimeSeconds(time_t now)
+time_t GetBootTime(void)
 {
-    time_t boot_time = 0;
+    static time_t boot_time = 0;
     errno = 0;
 
+    if (boot_time == 0)
+    {
 #if defined(BOOT_TIME_WITH_SYSINFO)         // Most GNU, Linux platforms
 
-    struct sysinfo s;
-    if (sysinfo(&s) == 0)
-    {
-       // Don't return yet, sanity checking below
-       boot_time = now - s.uptime;
-    }
+        time_t now = time(NULL);
```

Comment:
You're right, but I'm not arguing this.

What I'm commenting about is that `boot_time` is *static*, which means that
a long running process will have the wrong time for all its lifetime.

On Fri, Jun 19, 2015 at 12:17 PM, Kristian Amlie <notifications@github.com>
wrote:

> In libenv/sysinfo.c
> <https://github.com/cfengine/core/pull/2246#discussion_r32817172>:
>
> >  #if defined(BOOT_TIME_WITH_SYSINFO)         // Most GNU, Linux platforms
> >
> > -    struct sysinfo s;
> > -    if (sysinfo(&s) == 0)
> > -    {
> > -       // Don't return yet, sanity checking below
> > -       boot_time = now - s.uptime;
> > -    }
> > +        time_t now = time(NULL);
>
> In that case we get the same issue as before the fix, boot time will be
> zero (or another constant value), and processes may get the same start time
> unintentionally. Only after the time is set do we get unique start times.
>
> In theory you could defeat this mechanism by setting the time backwards,
> but you'd have to be *insanely* unlucky to both have the process start at
> the same second that the previous one started *and* get the same PID.
> This is practically impossible unless your intention is to trigger it and
> you try over and over. Certainly it is much more likely that the original
> issue described in https://dev.cfengine.com/issues/7075 will happen using
> the old method, than triggering what I just described with the new method.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cfengine/core/pull/2246/files#r32817172>.
>